src/Bicep.Core/Emit/TemplateEmitter.cs

```diff
@@ -89,7 +99,9 @@ public TemplateEmitter(SemanticModel model, EmitterSettings settings)
         /// <param name="writer">The json writer to write the template</param>
         public EmitResult Emit(JsonTextWriter writer) => this.EmitOrFail(() =>
         {
-            new TemplateWriter(this.model, this.settings).Write(writer);
+            var emitter = new TemplateWriter(this.model, this.settings);
```

Comment:
How about passing a `this.sourceMap` (need to change it to IDictionary and initialize it in the constructor) as a TemplateWriter constructor parameter? This would eliminate the need for the `this.sourceMap = ...` assignments in the `Emit` methods.

Reply:
I had to revert these changes with the record type updates to source map to avoid messing code due to instantiating the ImmutableArray members. I believe it also makes sense for the TemplateWriter to instantiate/generate the source map when it does the final processing for the mappings. Ideally, Write() would return the source map directly but that would return updating ITemplateWriter when the other implementation are not concerned with the source map.

Another option could be to add a callback method to ITemplateWriter.Write() that could do the `this.sourceMap = emitter.SourceMap` call, though that doesn't reduce the code much.